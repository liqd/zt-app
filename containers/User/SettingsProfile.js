import React from 'react'
import { Alert, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { Formik } from 'formik'
import * as yup from 'yup'

import API from '../../BaseApi'
import { ButtonSubmit } from '../../components/ButtonSubmit'
import { TextInputFormField } from '../../components/formFields'
import { Header } from '../../components/Header'
import { AvaterImageAddButton } from '../../components/imageFormField'
import { ListContainer, ListItem } from '../../components/List'

import { styles } from './SettingsProfile.styles'

export const SettingsProfile = props => {

  const {userName, userImage} = props.route.params

  const userNameValidationSchema = yup.object().shape({
    username: yup
      .string()
      .max(60, 'Required. 60 characters or fewer. Letters, digits, spaces and @/./+/-/_ only.')
  })

  const makeFormData = (values) => {
    // do not send null image on create
    if ('image' in values && values.image == null) {
      delete values['image']
    }
    let formData = new FormData()
    for (let key in values) {
      Array.isArray(values[key])
        ? values[key].forEach((value) => formData.append(key, value))
        : formData.append(key, values[key])
    }
    return formData
  }

  const handleSubmit = (values) => {
    const formData = makeFormData(values)
    AsyncStorage.getItem('authToken')
      .then((token) => {
        return API.editUser(formData, token)
      })
      .then((response) => {
        const {statusCode, data} = response
        if (statusCode == 200) {
          Alert.alert('You\'re profile has been updated')
          props.navigation.navigate({
            name: 'SettingsOverview',
            params: {
              userName: values.username,
              userImage: values.image.uri
            },
            merge: true,
          })
        } else {
          const errorMessage = 'That did not work.'
          let errorDetail
          if (statusCode==403) {
            errorDetail = data.detail
          } else if (statusCode == 400) {
            errorDetail = 'Bad request'
          }
          Alert.alert(errorMessage, errorDetail)
        }
      })
  }

  return (
    <SafeAreaView style={styles.container}>
      <Formik
        validationSchema={userNameValidationSchema}
        initialValues={
          {username: userName},
          {image: userImage}
        }
        onSubmit={values => handleSubmit(values)}
        validateOnMount={true}
      >
        {({
          handleChange,
          handleBlur,
          handleSubmit,
          values,
          errors,
          touched,
          isValid,
          setFieldValue
        }) => (
          <>
            <View>
              <Header transparent={true} navigation={props.navigation} />
              <ListContainer
                title='Edit Profile'>
                <ListItem>
                  <AvaterImageAddButton
                    name='image'
                    values={values.image}
                    field='Change profile picture'
                    fieldPreview='Profile picture'
                    labelText='change profile image'
                    title='profile image'
                    onSetImage={(img) => {
                      setFieldValue('image', img)
                    }}
                    imgSource={{uri: userImage}}
                    image={values.image ? values.image.uri : userImage}
                  />
                </ListItem>
                <TextInputFormField
                  username='username'
                  field='Username'
                  value={values.username}
                  placeholder={userName}
                  returnKeyType='next'
                  returnKeyLabel='next'
                  onChangeText={handleChange('username')}
                  onBlur={handleBlur('username')}
                  error={errors.username}
                  touched={touched.username}
                  textInputContainer={styles.textInputList}
                />
              </ListContainer>
            </View>
            <ButtonSubmit
              title='Save'
              onPress={handleSubmit}
              disabled={!isValid}
            />
          </>
        )}
      </Formik>
    </SafeAreaView>
  )
}
