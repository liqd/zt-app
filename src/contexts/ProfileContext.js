import React, { createContext, useState } from 'react'

export const ProfileContext = createContext()

export const ProfileProvider = (props) => {
  const [profileContext, setProfileContext] = useState()

  return (
    <ProfileContext.Provider value={[profileContext, setProfileContext]}>
      {props.children}
    </ProfileContext.Provider>
  )
}
