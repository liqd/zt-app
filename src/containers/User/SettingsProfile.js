import React, { useContext } from 'react'
import { Alert } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { Formik } from 'formik'
import * as yup from 'yup'

import API from '../../BaseApi'
import { ButtonAvatar } from '../../components/ButtonAvatar'
import { TextInputFormField } from '../../components/formFields'
import { Header } from '../../components/Header'
import { KeyboardScrollView } from '../../components/KeyboardScrollView'
import { ListContainer, ListItem } from '../../components/List'
import { StatusBarStyled } from '../../components/StatusBarStyled'
import { ProfileContext } from '../../contexts/ProfileContext'

import { styles } from './SettingsProfile.styles'

export const SettingsProfile = props => {
  const [profileContext, setProfileContext] = useContext(ProfileContext)

  const userNameValidationSchema = yup.object().shape({
    username: yup
      .string()
      .max(60, 'Required. 60 characters or fewer. Letters, digits, spaces and @/./+/-/_ only.')
  })

  const makeFormData = (values) => {
    // do not send image data if not changed
    if (!values['user_image']) {
      delete values['user_image']
    } else if (values['user_image'] === profileContext?.newUserImage) {
      delete values['user_image']
    } else if (values['user_image'] !== profileContext?.newUserImage) {
      // instead of setFieldValue (Formik), assigning "manually" here
      values['user_image'] = profileContext.newUserImage
      setProfileContext({
        ...profileContext,
        userImage: profileContext.newUserImage,
        newUserImage: null
      })
    }

    // update profile context if username changed
    if (values['username'] !== profileContext?.userName) {
      setProfileContext({...profileContext, userName: values['username']})
    }

    let formData = new FormData()
    for (let key in values) {
      Array.isArray(values[key])
        ? values[key].forEach((value) => formData.append(key, value))
        : formData.append(key, values[key])
    }
    return formData
  }

  const handleSubmit = (values) => {
    const formData = makeFormData(values)
    return API.editUser(formData)
      .then((response) => {
        const {statusCode, data} = response
        if (statusCode == 200) {
          Alert.alert('You\'re profile has been updated')
          props.navigation.navigate('SettingsOverview')
        } else {
          const errorMessage = 'That did not work.'
          let errorDetail
          if (statusCode==403) {
            errorDetail = data.detail
          } else if (statusCode == 400) {
            errorDetail = 'Bad request'
          }
          Alert.alert(errorMessage, errorDetail)
        }
      })
  }

  const toProfileSettingsAvatar = () => {
    props.navigation.navigate('SettingsProfileAvatar')
  }

  return (
    <SafeAreaView style={styles.container}>
      <StatusBarStyled />
      <Formik
        validationSchema={userNameValidationSchema}
        initialValues={{
          username: profileContext.userName,
          user_image: profileContext.userImage
        }}
        onSubmit={values => handleSubmit(values)}
        validateOnMount={true}
      >
        {({
          handleChange,
          handleBlur,
          handleSubmit,
          values,
          errors,
          touched,
          isValid
        }) => (
          <>
            <Header transparent={true} navigation={props.navigation} />
            <KeyboardScrollView
              handleSubmit={handleSubmit}
              isValid={isValid}
              buttonText='Save'
            >
              <ListContainer
                title='Edit Profile'>
                <ListItem>
                  <ButtonAvatar
                    imgSource={{uri:
                      profileContext?.newUserImage?.uri ||
                      profileContext?.userImage?.uri ||
                      profileContext?.userImageFallback?.uri
                    }}
                    a11yLabelText="Change profile picture"
                    a11yHintText="Click to navigate to change or add your profile image"
                    avatarStyles={styles.avatarStyles}
                    onPress={() => toProfileSettingsAvatar()}
                  >
                    {profileContext.userImage
                      ? 'Change profile picture'
                      : 'Add profile picture'
                    }
                  </ButtonAvatar>
                </ListItem>
                <TextInputFormField
                  username='username'
                  field='Username'
                  value={values.username}
                  placeholder={profileContext?.userName}
                  returnKeyType='next'
                  returnKeyLabel='next'
                  onChangeText={handleChange('username')}
                  onBlur={handleBlur('username')}
                  error={errors.username}
                  touched={touched.username}
                  textInputContainer={styles.textInputList}
                />
              </ListContainer>
            </KeyboardScrollView>
          </>
        )}
      </Formik>
    </SafeAreaView>
  )
}
