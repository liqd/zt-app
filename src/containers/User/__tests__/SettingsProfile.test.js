import React from 'react'
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react-native'

import '@testing-library/jest-native/extend-expect'

import API from '../../../BaseApi'
import { ProfileContext } from '../../../contexts/ProfileContext'
import { testUser } from '../../../tests/TestData'
import { SettingsProfile } from '../SettingsProfile.js'

test('Test SettingsOverview Snapshot no user image', async () => {
  const { toJSON, queryByText } = render(
    <ProfileContext.Provider value={[{
      userId: testUser.pk,
      userName: testUser.username,
      userImage: null
    }]}>
      <SettingsProfile />
    </ProfileContext.Provider>
  )
  await waitFor(() => {
    expect(queryByText(/Username/)).toBeTruthy()
  })
  expect(toJSON()).toMatchSnapshot()
})

test('Test SettingsOverview, change username', async () => {
  render(
    <ProfileContext.Provider value={[{
      userId: testUser.pk,
      userName: testUser.username,
      userImage: null
    }]}>
      <SettingsProfile />
    </ProfileContext.Provider>
  )
  const nameInput = await screen.findByPlaceholderText(testUser.username)

  await act(async () => {
    fireEvent.changeText(nameInput, 'newUserName')
  })
  expect(nameInput).toHaveProp('value', 'newUserName')
})

xtest('Test SettingsOverview, send updated username', async () => {
  const mockAppend = jest.fn()
  function FormDataMock() {
    this.append = mockAppend
  }
  global.FormData = FormDataMock

  API.editUser = jest.fn(() =>
    Promise.resolve({
      statusCode: 200
    })
  )
  render(
    <ProfileContext.Provider value={[{
      userId: testUser.pk,
      userName: testUser.username,
      userImage: null
    }]}>
      <SettingsProfile />
    </ProfileContext.Provider>
  )

  const nameInput = await screen.findByPlaceholderText(testUser.username)
  const submitButton = screen.getByText('Save')
  await act(async () => {
    fireEvent.changeText(nameInput, 'newUserName')
  })
  expect(nameInput).toHaveProp('value', 'newUserName')
  await act(async () => {
    fireEvent.press(submitButton)
  })
  expect(API.editUser).toHaveBeenCalled()
  expect(mockAppend).toHaveBeenCalledWith('username', 'newUserName')
})
