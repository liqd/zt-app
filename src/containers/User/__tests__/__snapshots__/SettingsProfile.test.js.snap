// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SettingsOverview Snapshot no user image 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={Object {}}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingBottom": 20,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 2,
            },
            undefined,
            false,
          ]
        }
        testID="RNE_BUTTON_WRAPPER"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#2089dc",
                "borderRadius": 2,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginVertical": 0,
                "padding": 8,
                "paddingHorizontal": 10,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 22,
                  },
                  undefined,
                  Object {
                    "fontFamily": "simple-line-icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              î˜…
            </Text>
            <Text
              style={
                Object {
                  "color": "#1a1a1a",
                  "fontSize": 19.125,
                  "marginHorizontal": 5,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Back
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "space-between",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "SourceSansPro_400Regular",
                  "fontSize": 18,
                },
                Object {
                  "backgroundColor": "#f2f2f2",
                  "borderBottomColor": "#d8d8d8",
                  "borderBottomWidth": 1,
                  "borderRadius": 0,
                  "fontFamily": "SourceSansPro_600SemiBold",
                  "fontSize": 19.125,
                  "paddingLeft": 15,
                  "paddingVertical": 15,
                },
              ]
            }
          >
            Edit Profile
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderBottomColor": "#d8d8d8",
                "borderBottomWidth": 1,
                "color": "#1a1a1a",
                "flexDirection": "row",
                "flexWrap": "wrap",
                "padding": 15,
                "width": "100%",
              }
            }
          >
            <View
              accessibilityHint="Click to navigate to change or add your profile image"
              accessibilityLabel="Change profile picture"
              accessibilityRole="button"
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#fff",
                  "display": "flex",
                  "flexDirection": "row",
                  "opacity": 1,
                }
              }
            >
              <Image
                accessibilityIgnoresInvertColors={true}
                source={
                  Object {
                    "uri": undefined,
                  }
                }
                style={
                  Array [
                    Object {
                      "backgroundColor": "#2d40cc",
                      "borderRadius": 100,
                      "height": 40,
                      "width": 40,
                    },
                    Object {
                      "marginRight": 20,
                    },
                  ]
                }
              />
              <Text
                style={
                  Object {
                    "fontFamily": "SourceSansPro_400Regular",
                    "fontSize": 18,
                  }
                }
              >
                Add profile picture
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "borderBottomColor": "#d8d8d8",
                "borderBottomWidth": 1,
                "paddingHorizontal": 15,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "SourceSansPro_400Regular",
                    "fontSize": 18,
                  },
                  Object {
                    "fontSize": 16.875,
                    "lineHeight": 20.9,
                    "paddingBottom": 5,
                    "paddingTop": 30,
                    "width": "100%",
                  },
                ]
              }
            >
              Username
            </Text>
            <TextInput
              accessibilityLabel="Username"
              accessible={true}
              field="Username"
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="test-user"
              returnKeyLabel="next"
              returnKeyType="next"
              style={
                Object {
                  "borderBottomColor": "#d8d8d8",
                  "borderBottomWidth": 1,
                  "fontSize": 19.125,
                  "height": 40,
                }
              }
              textInputContainer={
                Object {
                  "backgroundColor": "#fff",
                  "borderBottomColor": "#d8d8d8",
                  "borderBottomWidth": 1,
                  "paddingHorizontal": 15,
                }
              }
              username="username"
              value="test-user"
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityLabel="Save"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2d40cc",
              "borderColor": "#2089dc",
              "borderRadius": 0,
              "borderWidth": 0,
              "flexDirection": "row",
              "fontSize": 18,
              "height": 113.39000000000001,
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Save
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
